uvicorn/main.py

```diff
@@ -271,10 +271,18 @@ def run(app, **kwargs):
         sock = config.bind_socket()
         supervisor = StatReload(config)
         supervisor.run(server.run, sockets=[sock])
+        sock.close()
     elif config.workers > 1:
-        sock = config.bind_socket()
         supervisor = Multiprocess(config)
-        supervisor.run(server.run, sockets=[sock])
+        if config.uds is not None:
+            sock = config.bind_unix_socket()
+            supervisor.run(server.run, sockets=[sock])
+            sock.close()
+            os.remove(config.uds)
```

Comment:
Curious: Why do we remove the UDS socket here?

Reply:
From `man 7 Unix`

> Binding  to  a  socket with a filename creates a socket in the filesystem that must be deleted by the caller when it is no longer needed (using unlink(2)).  The usual UNIX close-behind semantics apply; the socket can be unlinked at any time and will be finally removed from the filesystem when the last reference to it is closed.

I think there is another problem. If our process crashes, `os.remove()` will not be called and we’ll have a dangling socket. It seems that we need to check the socket file before `bind()`.


----
Why haven’t encountered this problem when use one worker and unix socket mode?

Because we use the `create_unix_server`

https://github.com/encode/uvicorn/blob/e0a77de6baf436d182868fa280cd5e2f0d66501c/uvicorn/main.py#L363-L373

`create_unix_server` will unlink the socket file before it starts.

https://github.com/python/cpython/blob/0d3fe8ae4961bf551e7d5e42559e2ede1a08fd7c/Lib/asyncio/unix_events.py#L283-L293